server/src/controllers/GameControllers/snake.js

```diff
@@ -0,0 +1,202 @@
+    return false;
+  }
+
+  /**
+   * Checks collision of the head with it's rival's body.
+   */
+  hasCollidedWithRival() {
+    const headTip = this.getHeadTip();
+    return isPointInBody(this.rivalBody, headTip);
+  }
+
+  /**
+   * We are drawing arcs on each cooridnate point in the snake body, so to check collision we are getting the most ahead part of the snake head.
+   */
+  getHeadTip() {
+    const arcRadius = SNAKE_CONSTANTS.ARC_RADIUS;
+    const { x, y } = this.head;
+    let headTip;
+    switch (this.direction) {
+      case 'down': {
+        headTip = { x, y: y + arcRadius };
+        break;
+      }
+      case 'up': {
+        headTip = { x, y: y - arcRadius };
+        break;
+      }
+      case 'left': {
+        headTip = { x: x - arcRadius, y };
+        break;
+      }
+      case 'right': {
+        headTip = { x: x + arcRadius, y };
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+    return headTip;
+  }
+
+  changeDirection(key) {
+    // eslint-disable-next-line default-case
+    switch (key) {
+      case 'ArrowLeft': {
+        this.handleDirectionChange('left', 'right', -2, 0);
+        break;
+      }
+      case 'ArrowRight': {
+        this.handleDirectionChange('right', 'left', 2, 0);
+        break;
+      }
+      case 'ArrowUp': {
+        this.handleDirectionChange('up', 'down', 0, -2);
+        break;
+      }
+      case 'ArrowDown': {
+        this.handleDirectionChange('down', 'up', 0, 2);
+        break;
+      }
+    }
+  }
+
+  /**
+   * @param  {String} direction the key press direction from the user.
+   * @param  {String} oppositeDirection the opposite direction of the user clicked arrow, so we can handle some edge cases in direction change.
+   * @param  {Integer} xVelocity the velocity that will be given on the direction change on xAxis.
+   * @param  {Integer} yVelocity the velocity which will be given on the direction change on yAxis.
+   */
+  handleDirectionChange(direction, oppositeDirection, xVelocity, yVelocity) {
+    // the direction is already on the same direction or opposite direction snake will not change the directioin
+    if (this.direction === direction || this.direction === oppositeDirection) {
+      return false;
+    }
+    // if snake is taking a u-turn then below condition will be true, and it does not collide with itself, so we are giving it a smooth handling of you turn.
+    if (this.previousDirection !== direction && this.previousDirection === oppositeDirection) {
+      this.makeSmoothUTurn();
+    }
+    this.xAxisVelocity = xVelocity;
+    this.yAxisVelocity = +yVelocity;
+    this.previousDirection = this.direction;
+    this.direction = direction;
+    return true;
+  }
+
+  /**
+   * This function ensures smooth u-turn on board(UI) when user takes 180 degree turn.
+   */
+  makeSmoothUTurn() {
+    // @TODO:Refactor this code in future version, either user lodash or move it into some utility file.
+    this.moveSnakeOneStep();
```

Comment:
I need to give it some thought, Right now if the user takes u-turn really fast, then snake collides with itself because of the logic. I have put a TODO here so I can handle it later. Probably, when we are done with the user flow.